rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2973 @@
+```
+
+
+# Motivation
+[motivation]: #motivation
+
+We foresee a range of practical consequences of fulfilling these goals, including [making unsafe Rust safer](#making-unsafe-rust-safer), providing [safer initialization primitives](#safer-initialization-primitives), and [generic atomics](#atomict). We explore these use-cases, and more, in this section.
+
+A *comprehensive* approach to safe transmutation provides benefits beyond the mere act of transmutation; namely:
+ - [authoritatively codifies language layout guarantees](#codifying-language-layout-guarantees)
+ - [allows crate authors to codify their types' layout stability guarantees](#expressing-library-layout-guarantees)
+ - [allows crate authors to codify their abstractions' layout requirements](#expressing-layout-requirements)
+
+## Codifying Language Layout Guarantees
+Documentation of Rust's layout guarantees for a type are often spread across countless issues, pull requests, RFCs and various official resources. It can be very difficult to get a straight answer. When transmutation is involved, users must reason about the *combined* layout properties of the source and destination types.
+
+This RFC proposes mechanisms that programmers will use to confidently answer such questions—by checking whether the `TransmuteFrom` and `TransmuteInto` traits are implemented, or (equivalently) by checking whether the `can_transmute` predicate (a `const fn`) is satisfied.
+
+## Expressing Library Layout Guarantees
+There is no canonical way for crate authors to declare the SemVer layout guarantees of their types. Crate authors currently must state their layout guarantees using prose in their documentation. In contrast to structural stability (e.g., the declared visibility of fields), layout stability is expressed extra-linguistically.
+
+This isn't satisfactory: guarantees expressed in prose outside of the Rust programming language are guarantees that cannot be reasoned about *inside* the language. Whereas `rustc` can dutifully deny programmers access to private fields, it is unable to prevent programmers from making unfounded expectations of types' in-memory layouts.
+
+This RFC proposes simple-but-powerful [mechanisms][stability] for declaring layout stability guarantees.
+
+## Expressing Layout Requirements
+Similarly, there is no canonical way for crate authors to declare the layout requirements of generic abstractions over types that have certain layout properties. 
+
+For instance, a common bit-packing technique involves abusing the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.
```

Comment:
```suggestion
For instance, a common bit-packing technique involves using the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.
```

Tiny nit, but using a property for something it may not have been intended for is still `use`, not `abuse`.